forc-pkg/src/lock.rs

```diff
@@ -296,30 +296,25 @@ fn pkg_dep_line(
 type ParsedPkgLine<'a> = (Option<fuel_tx::Salt>, Option<&'a str>, &'a str);
 // Parse the given `PkgDepLine` into its dependency name and unique string segments.
 //
-// I.e. given "(<salt>) (<dep_name>) <name> <source>", returns ("<salt>", "<dep_name>", "<name> <source>").
+// I.e. given "(<dep_name>) <name> <source> (<salt>)", returns ("<salt>", "<dep_name>", "<name> <source>").
 //
 // Note that <source> may not appear in the case it is not required for disambiguation.
 fn parse_pkg_dep_line(pkg_dep_line: &str) -> anyhow::Result<ParsedPkgLine> {
     let s = pkg_dep_line.trim();
 
-    // Check for the open bracket.
-    let salt = if !s.starts_with('(') {
-        None
-    } else {
-        // If we have the open bracket, grab everything until the closing bracket.
-        let s = &s["(".len()..];
-        let mut iter = s.split(')');
-        let dep_salt = iter
-            .next()
-            .ok_or_else(|| anyhow!("missing closing parenthesis"))?;
-        // It is possible that we don't have a salt but a dep_name so we do not want to error out
-        // instead we want to return Some(Salt) if the str we grabbed is corresponding to a `Salt`.
-        fuel_tx::Salt::from_str(dep_salt).ok()
-    };
-
     // Check for the open bracket.
     if !s.starts_with('(') {
-        return Ok((salt, None, s));
+        let mut iter = s.split('(');
+        let pkg_str = iter
+            .next()
+            .ok_or_else(|| anyhow!("missing pkg string"))?
+            .trim();
+        let salt = iter
+            .next()
+            .map(|s| s.trim())
+            .map(|s| &s[..s.len() - 1])
+            .and_then(|s| fuel_tx::Salt::from_str(s).ok());
```

Comment:
Looking a little closer I see there's some code duplication where we check for and parse the salt here, as well as later on in the function.

The purpose of this condition where we check `!s.starts_with('(')` is only to check for the `dep_name`.

Rather than parsing the `pkg_str` and `salt` in this condition and returning early, how about we instead do something like this:

```rust
    let (dep_name, s) = match s.starts_with('(') {
        false => (None, s),
        true => {
            // If we have the open bracket, grab everything until the closing bracket.
            let s = &s["(".len()..];
            let mut iter = s.split(')');
            let dep_name = iter
                .next()
                .ok_or_else(|| anyhow!("missing closing parenthesis"))?;
            // The rest is the unique package string and possibly the salt.
            let s = &s[dep_name.len() + ")".len()..];
            (Some(dep_name), s)
        }
    };
```

This should help to avoid the early return and the code duplication related to salt parsing.

Reply:
Yeah, that `.ok()` part was there for handling `salt` prepended to the `pkg_dep_line` (to distinguish from `dep_name`) since we moved it to the end there is no need for it anymore. Also removed the duplication as you suggested. Thanks!